Playwright: write test for Signup: WPCC path.

worldomonation opened the pull request:
#### Changes proposed in this Pull Request

This PR adds coverage of the WPCC signup path.

Key changes:
- new test that signs up for a service using WPCC, then verifies the WordPress.com account is able to interact with some features in Calypso.
- removal of deprecated or migrated test suites from Selenium version of `Signup` suite.
- slight changes to existing Signup suites for suite names and Mailosaur inbox IDs.


#### Testing instructions

- [ ] pre-release tests
  - [ ] pre-release
- [x] pr tests
  - [x] mobile
  - [x] desktop


Related to #

worldomonation committed:
Develop new tests for WPCC endpoint.
- create account via CrowdSignal.
- interact in the account with various features.
- new methods and updated selectors for relevant page objects.

Rename or add details to other Signup specs.
- speicify the endpoint that is being tested.

worldomonation committed:
Update selector used for SupportComponent to not overlap with Reader stream's search input.

Fix desktop signup path for WPCC.

Use CrowsSignal Oauth2 endpoint but with hostname swapped out for the calypsoBaseURL under test.

dpasque commented on test/e2e/specs/specs-playwright/wp-signup__wpcc.ts:
```diff
@@ -0,0 +1,135 @@
+				await readerPage.search( 'e2eflowtestingnotifications' );
+			} );
+
+			it( 'Read first post', async function () {
+				await readerPage.visitPost( { index: 1 } );
+			} );
+
+			it( 'Follow site', async function () {
+				await readerPage.followSite();
+			} );
+
+			it( 'Navigate back to search listing', async function () {
+				await page.goBack();
+			} );
```
I'm not entirely sure of my own answer to this, but wanted to throw it out there: do you think we need these test cases here? Or could we add them to the reader spec?

I'm just keeping the goal in mind to prevent as much bloat in the pre-release tests as possible, and I _think_ as long as we get in the reader and can search, I feel okay then offloading the other tests to the reader spec.

What do you think?

worldomonation replied:
+1 to a good point. 

Tests for non-Signup related features were included to behave like what a user would do; sign up then use the account for things. Manly, I was looking to test the following:

- did the signup process succeed?
- if it did, did the account get set up correctly?
- is the account able to perform actions that it's eligible to?

Which is why I have the reader piece in this spec and magic link in another and site launch in another.

With that said, I am open to having some tests be trimmed down to only test the signup flow and perhaps this one (WPCC) is a good candidate.

dpasque replied:
Yeah, to be honest I'm really torn. You make such a good point - really the goal of this level of testing is imitate real flows! What would a user want to do once they had signed up in this way? These kind of tests get at that. But also, we're always trying to keep this level torn. It's a hard tradeoff! 

@WunderBart - got any perspective?

I think I'm also down with the approach of _starting_ more comprehensive and in-depth, watching what that looks like, and then pruning back as needed.

WunderBart replied:
This is a tough one! 😄 In my experience, it's always best to draw a solid line between testing functionalities that should work independently from each other. For example, does logging in with Magic Link need to be preceded with account signup, or should it work independently for every registered user? If the latter, then we should have a separate suite (file) for it as we don't want any failing signup spec blocking the Magic Link login one. The same thing with the `Follow site` test, but from the opposite perspective now - if it fails, does it mean that the signup flow is broken, or that there's something wrong i.e. with follow button only? If it's the button, it shouldn't break the whole signup suite, IMO. Does that make sense?

worldomonation replied:
> if it fails, does it mean that the signup flow is broken, or that there's something wrong i.e. with follow button only?

This is one of the challenges we face and the have to make a calculated tradeoff of the risks.

From my understanding, we apply different adjustments to accounts based on when they were created. There was an example recently of a screen being enabled for all users (account closure?). Prior to this, my assumption is that the 'new' screen was enabled only for newer accounts. 

I am unaware of any other 'features' being turned on only for newer accounts but this is only because I haven't followed these sorts of experiments actively. Given the example with the Close Account flow above, I think we should assume that newly registered users get a slightly different experience than older accounts.

_With all that said, however_ - I can totally see the argument for why tests for features such as Magic Link or Reader Stream should _probably_ be moved out of Signup tests.

WunderBart replied:
> From my understanding, we apply different adjustments to accounts based on when they were created. There was an example recently of a screen being enabled for all users (account closure?). Prior to this, my assumption is that the 'new' screen was enabled only for newer accounts.

IIUC, it sounds to me like the new account closure screen would naturally become a part of the signup test suite, as we create and close accounts at the time of running the test - is that correct?

Anyway, I think we're on the same page actually, where only features that are available for newly registered users should be a part of the signup suite. That's why moving out Magic Link or Reader specs would make sense!

worldomonation replied:
> IIUC, it sounds to me like the new account closure screen would naturally become a part of the signup test suite, as we create and close accounts at the time of running the test - is that correct?

In the case of _Close Accounts_, yes it would have become an inherent part of the flow, but I assume there may be other areas of Calypso where the experience may differ between new and old users.

> Anyway, I think we're on the same page actually

Sounds good, I can start by moving out the extra tests out of this spec. I'll spin up a new PR for the move.

worldomonation replied:
I've moved out the Support and Reader specs out of this test.

Support has a new PR #56622.
Reader will need to be carefully considered - Following a site alters the state, so it would need to likely run in isolation as a Pre-Release test, not as part of every commit.

dpasque commented on test/e2e/specs/specs-playwright/wp-signup__wpcc.ts:
```diff
@@ -0,0 +1,135 @@
+			it( 'Navigate to CrowdSignal and click to try for free', async function () {
+				//  No POM for this page.
+				await page.goto( 'https://crowdsignal.com/' );
+				await Promise.all( [ page.waitForNavigation(), page.click( 'a:text("Try it free")' ) ] );
+			} );
+
+			it( 'Create a new WordPress.com account', async function () {
+				const userSignupPage = new UserSignupPage( page );
+				await userSignupPage.signupWPCC( email, signupPassword );
```
I love the crowdsignal integration, but I also think that this means that we can't run this test against a calypso.live, or our `wpcalypso.wordpress.com` endpoint because this will always go to wordpress.com.

I think we might have two options here:
1. Can we just jump directly to a WPCC auth route on the `calypsoBaseUrl`, and just bypass the crowdsignal piece of this? As cool as crowdsignal integration is, it might not also be a bad idea to limit dependencies during the pre-release execution time
2. We could probably intercept the link, parse it, swap out the host, and then go to right there. That would keep the crowdsignal integration intact, but allow us to use the right host client!

worldomonation replied:
> As cool as crowdsignal integration is, it might not also be a bad idea to limit dependencies during the pre-release exec

Yeah, that's an excellent critique. Currently, this _likely_ ends up running against production, which would be...pointless.

worldomonation replied:
There's this link (as per FG) that could serve as the direct link to the WPCC auth route: https://wordpress.com/start/crowdsignal/oauth2-name?ref=oauth2&oauth2_redirect=https%3A%2F%2Fpublic-api.wordpress.com%2Foauth2%2Fauthorize%2F%3Fclient_id%3D978%26response_type%3Dcode%26blog_id%3D0%26state%3D34604bf9bcb69d07819b97c65b744925c061ed944e4d30623b9b8cebf1a4ce34%26redirect_uri%3Dhttps%253A%252F%252Fapp.crowdsignal.com%252Fconnect%253Fsource%253Dsignup%2526action%253Drequest_access_token%26wpcom_connect%3D1%26jetpack-code%26jetpack-user-id%3D0%26action%3Doauth2-login&oauth2_client_id=978

dpasque replied:
oh wowsers! Generating all those query params on the fly might be a bit onerous to maintain... I think we could just start by looking at the link and swapping out the hostname. I think we could run it for a bit a see if the crowdsignal integration ends up causing flakiness?

worldomonation replied:
I verified that `wpcalypso.wordpress.com/start/crowdsignal/...` works and loads the `wpcalypso` version of the webapp. 

@bsessions85 @kwight for lack of knowing who else to ask - do you see an issue with the approach used to test WPCC signup in this PR? 

The script is hitting the WPCC endpoint for CrowdSignal (above), fill out information and submit. This creates an account for the user, and the script goes on to try a few more things before closing the account.

dpasque reviewed (commented):
Overall a really clean, beautiful migration, nice work!

I had a couple (not firm) questions on test designs. Want to let me know your 0.02?

worldomonation commented:
Taking back the PR for some more updates.

worldomonation commented on packages/calypso-e2e/src/lib/components/support-component.ts:
```diff
@@ -6,7 +6,7 @@ const selectors = {
 	supportPopoverButton: '.inline-help__button',
 	supportPopover: '.inline-help__popover',
-	searchInput: '[aria-label="Search"]',
+	searchInput: '.inline-help__search input[aria-label="Search"]',
```
This is to eliminate conflict with Reader page's search field.

worldomonation commented on packages/calypso-e2e/src/lib/flows/gutenboarding-flow.ts:
```diff
@@ -81,7 +92,7 @@ export class GutenboardingFlow {
 	 * @param {string} title Title of the site.
 	 */
-	async enterSiteTitle( title: string ) {
+	async enterSiteTitle( title: string ): Promise< void > {
```
Linter fix.

worldomonation commented on packages/calypso-e2e/src/lib/pages/user-signup-page.ts:
```diff
@@ -47,4 +57,39 @@ export class UserSignupPage {
+	 * @param {string} email Email address of the new user.
+	 * @param {string} password Password of the new user.
+	 */
+	async signupWPCC( email: string, password: string ): Promise< void > {
```
This implementation of WPCC login is based on CrowdSignal WPCC.

WunderBart commented on packages/calypso-e2e/src/lib/pages/reader-page.ts:
```diff
@@ -35,6 +38,25 @@ export class ReaderPage {
+	 */
+	async search( keyword: string ): Promise< void > {
+		await Promise.all( [
+			this.page.waitForNavigation(),
```
Does this have any effect here? I'm asking because it doesn't seem like it's redirecting, only setting query params to the URL. Maybe using the `waitForResponse` would be a better match? Something like:

```suggestion
			this.page.waitForResponse( /read\/search/ ),
```

worldomonation replied:
Playwright treats this as a navigation. See the following snippet generated with `DEBUG=pw:api` set:

```
  pw:api   navigated to "https://wordpress.com/read/search?q=&focus=1" +164ms
  pw:api   "load" event fired +281ms
  pw:api   "networkidle" event fired +4s
  pw:api   navigated to "https://wordpress.com/read/search?q=Test+String&focus=1" +94ms
```

WunderBart replied:
Thanks for the snippet. Looks like every change of the URL is considered a navigation change, then! Good to know.

WunderBart commented on packages/calypso-e2e/src/lib/flows/gutenboarding-flow.ts:
```diff
@@ -68,10 +70,19 @@ export class GutenboardingFlow {
+	async clickWpLogo(): Promise< void > {
+		await Promise.all( [ this.page.waitForNavigation(), this.page.click( selectors.wpLogo ) ] );
+
+		await this.page.waitForLoadState( 'networkidle' );
```
I'd make sure this is really necessary here! In my experience, it can lead to unnecessarily long wait times as there is something that's still loading that shouldn't really stop the test from moving on.

worldomonation replied:
Originally, this was a `waitForLoadState('load')` but that was not enough of a wait; the pending navigations ended up swallowing the click on the navbar.

worldomonation replied:
I tested some more and the results seem to indicate the standard `load` state would suffice, and that's already a part of the `NavbarComponent.clickMe` method that constitutes the `CloseAccountFlow`.

WunderBart commented on packages/calypso-e2e/src/lib/pages/user-signup-page.ts:
```diff
@@ -47,4 +57,39 @@ export class UserSignupPage {
+		await this.page.fill( selectors.emailInput, email );
+		await this.page.fill( selectors.passwordInput, password );
+
+		if ( target === 'mobile' ) {
+			await Promise.all( [
+				this.page.waitForNavigation(),
+				this.page.click( selectors.submitButton ),
+			] );
+		} else {
+			await Promise.all( [
+				this.page.waitForNavigation(),
+				this.page.click( selectors.createWPCOMAccountButton ),
+			] );
+		}
```
Do you think it would be better to make the submit button selector depend on the target device?

```js
const selectors = {
  // ...
  submitButton: target === 'mobile'
    ? 'button:text("Create a WordPress.com Account"):visible'
    : 'button[type="submit"]',
}
```

worldomonation replied:
Yes and no - in mobile, the user must first click on 'Create a WordPress.com account' button to reveal the form. The button with the same text then becomes a submit button.

However on desktop the 'Create a WordPress.com account' button is always visible and behaves as the submit button.

Here's an alternative solution I came up with while poking around in the DOM a bit more:

```
const selectors = {
   submitButton: button[type="submit"],
   createWPAccountButton: button:text("Create a WordPress.com Account"):visible
}
```

then, the relevant part of the `signupWPCC` method can simply call:

```
await Promise.all( [
	this.page.waitForNavigation(),
	this.page.click( selectors.submitButton ),
] );
```

worldomonation replied:
Fixed in c06160a60fea8e7e37c9965e404be4fc1c25e7bf.

WunderBart commented on test/e2e/specs/specs-playwright/wp-signup__wpcc.ts:
```diff
@@ -0,0 +1,139 @@
+				let calypsoBaseURL = DataHelper.getCalypsoURL();
+				calypsoBaseURL =
+					calypsoBaseURL +
+					'/start/crowdsignal/oauth2-name?ref=oauth2&oauth2_redirect=https%3A%2F%2Fpublic-api.wordpress.com%2Foauth2%2Fauthorize%2F%3Fclient_id%3D978%26response_type%3Dcode%26blog_id%3D0%26state%3D34604bf9bcb69d07819b97c65b744925c061ed944e4d30623b9b8cebf1a4ce34%26redirect_uri%3Dhttps%253A%252F%252Fapp.crowdsignal.com%252Fconnect%253Fsource%253Dsignup%2526action%253Drequest_access_token%26wpcom_connect ...
```
If this is a constant value, should we make it a part of the config object? This could then become

```js
calypsoBaseURL += DataHelper.config.get( 'WPCCAuthRoutePath' );
```

worldomonation replied:
Yeah that's probably a good way to go about it.

worldomonation replied:
Fixed in c06160a60fea8e7e37c9965e404be4fc1c25e7bf.

WunderBart reviewed (commented):
Just a few comments!

WunderBart commented on test/e2e/specs/specs-playwright/wp-signup__wpcc.ts:
```diff
@@ -0,0 +1,139 @@
+			} );
+		} );
+
+		describe( 'Delete user account and ensure no longer accessible', function () {
+			let loginPage: LoginPage;
+
+			it( 'Close account', async function () {
+				const closeAccountFlow = new CloseAccountFlow( page );
+				await closeAccountFlow.closeAccount();
+			} );
+
+			it( 'Navigate to WordPress.com Login page', async function () {
+				loginPage = new LoginPage( page );
+				await loginPage.visit();
+			} );
+
+			it( 'Ensure user is unable to log in', async function () {
+				await loginPage.login( { username: email, password: signupPassword } );
+				await page.waitForSelector( ':text("This account has been closed")' );
+			} );
+		} );
```
I've noticed that the setup (create username & password) and teardown (close account) steps are almost identical for every signup suite (free/paid/wpcc). IMO, it would be good to make those shareable between the suites for better consistency. For example, currently, only the wpcc suite's account closing is followed by ensuring that the account is closed. What do you think? cc @dpasque

worldomonation replied:
This following basic flow is indeed shared across all Signup tests:

```

			it( 'Close account', async function () {
				const closeAccountFlow = new CloseAccountFlow( page );
				await closeAccountFlow.closeAccount();
			} );
```

The two other steps are specific to WPCC Signup in that I am taking additional steps to verify the account closure indeed took place. In fact the `'Ensure user is unable to log in'` step does no exist anywhere else. I don't know if it's necessary to test that across every Signup test.

worldomonation replied:
As for the setup, I personally think it's better form to set up the values the test will use in the test itself instead of moving them behind a 'shared step' or something similar. Having a 'shared step' makes it difficult to follow the flow in my opinion.

Perhaps it would make sense to have another method in `DataHelper` which returns a generated email/password pair? That would keep the explicitness while removing one line of unnecessary setup.

eg.
```
[email, password] = DataHelper.getNewUserCredentials( {inboxId: inboxId, prefix: username} )
```

WunderBart replied:
> As for the setup, I personally think it's better form to set up the values the test will use in the test itself instead of moving them behind a 'shared step' or something similar. Having a 'shared step' makes it difficult to follow the flow in my opinion.

Makes sense, let's not share steps 👍 

> Perhaps it would make sense to have another method in DataHelper which returns a generated email/password pair?

The `getNewUserCredentials` helper sounds good to me - would be nice to have such a wrapper 👍

worldomonation replied:
@WunderBart Mind if this change is proposed in a separate PR?

WunderBart replied:
Absolutely 👍

worldomonation committed:
PR review comments from Bart:
- move WPCC path to a config value item.
- simplify signupWPCC method to eliminate branching.

Remove `networkidle`.

Fix the WPCC OAuth2 param value

worldomonation committed:
Remove non-essential tests from spec.
Re-add wpccAuthPath value into secrets file

worldomonation committed:
Add a wait

worldomonation committed:
Revert changes to files that have been split into its own PR: #56622.

WunderBart reviewed (approved):
🚀
